### ISSUE TITLE
float type has been changed to int incorrectly

### ISSUE THREAD

**Issue by AlirezaHaghi** [microsoft/pyright#11114]
**Describe the bug**
The type "int" is inferred incorrectly. Just copy and paste the snippet code.

**Code or Screenshots**
<img width="654" height="133" alt="Image" src="https://github.com/user-attachments/assets/d7621bca-3238-4737-b940-ab21453fa9a1" />

```python
def _normalize_prices(price: float | list[float] | None) -> list[float]:
    """Convert price input to a list of floats."""
    if price is None:
        return []
    result = [price] if isinstance(price, float) else price
    return result
```

**VS Code extension or command-line**
pylance: 2025.9.1

**Comment by erictraut** [comment]
Pyright's behavior here is correct. In the Python type system, the type annotation `float` implies `float | int`. See [this section of the typing spec](https://typing.python.org/en/latest/spec/special-types.html#special-cases-for-float-and-complex) for details. This means it's possible to pass an `int` value to your function without violating any typing rules. Your code doesn't guard against this case in its runtime `isinstance` check, so the narrowed type of `result` must include an `int`.

If you need convincing, try calling your function with an argument of `1`.

```python
print(_normalize_prices(1))
```

**Comment by AlirezaHaghi** [comment]
@erictraut 
It doesn't make sense at all.
>  the type annotation float implies float | int

This is trivial. A "float" can have an "int" value, but not vice versa.
Of course, it includes (and must consist of) the "int", but degenerating the float to the int is a BUG.
The fact that this method can accept an int is not controversial when compared to receiving a float.
pyright correctly identified another acceptable type is list[float] not list[int]

i can not understand why it narrowed down to int.
it's like you say:
`a : float` and infer that `b = 2*a` is an int.
this is absolutely a bug

FYI: I checked this with mypy. -> no error

**Comment by erictraut** [comment]
Your function's return type is annotated as `list[float]`. If you call your function with an `int` argument (as I've shown in my code snippet above), your function will not return a `list[float]` value. Instead, it returns the original integer value. That means there is a type-related bug in your code, so you should be surprised if a type checker does not report an error here.

Mypy does not catch this class of bug currently. That is a shortcoming of mypy.

**Comment by AlirezaHaghi** [comment]
@erictraut 
You know what it looks like?
Consider that you have a function like sin() and say the sin function at Pi/2 returns 1, and you implied that the type of sin is int.
"int" values are a subset of float values.
Even in this scenario, the type should be inferred as `float | int`, not just `int`.
If your function can work with floats, it should definitely work with integers. All well-known mathematical functions (even log(x) and exp(x)) return int values in some points.
please tell me: what is the type of log ans sin funcitons?
